### ISSUE TITLE
CxFlow occasionally fails to parse scan results and create bugtracker issues

### ISSUE THREAD

**Issue by y4ppieflu** [checkmarx-ltd/cx-flow#784]
### Description

Jira Bug Tracker sometimes fails to properly parse the scan results, thus the scan report is not getting handled and bugtracker issues are not created. Once the error occurs the first time, it then happens for all subsequent scans. This happens several times a day and restarting CxFlow resolves the issue for some time.

### Expected Behavior
When the scan is finished, scan results are parsed and bugtracker issues are created/updated.
### Actual Behavior
When the scan is finished, CxFlow fails to properly parse the scan results and throws an exception
### Reproduction
SCM: GitHub
BugTracker: Jira

Logs:
`2021-07-16 10:16:44.208  INFO 125534 --- [      flow-web1] c.c.f.s.AbstractVulnerabilityScanner      [H9csmv7i] : Checking if there is any existing scan for Project: 188
2021-07-16 10:16:44.215  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : No scans in the queue that are in progress
2021-07-16 10:16:44.215  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Creating scan...
2021-07-16 10:16:44.224  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Updating Source details for project Id 188
2021-07-16 10:16:44.918  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Scan will be Full Scan
2021-07-16 10:16:44.918  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Creating Scan for project Id 188
2021-07-16 10:16:44.932  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Scan created with Id 1020766 for project Id 188
2021-07-16 10:16:44.932 DEBUG 125534 --- [      flow-web1] c.c.f.s.BugTrackerEventTrigger            [H9csmv7i] : Bug tracker events were not triggered, because bug tracker type is 'JIRA'.
2021-07-16 10:17:04.955  INFO 125534 --- [      flow-web1] jsonLogger                                [H9csmv7i] : 
2021-07-16 10:17:04.956  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Creating report for xml Id 1020766
2021-07-16 10:17:04.973  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Report with Id 1459 created
2021-07-16 10:17:09.974  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Retrieving report status of report Id 1459
2021-07-16 10:17:10.988  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Retrieving report contents of report Id 1459 in XML format
2021-07-16 10:17:11.034  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Report downloaded for report Id 1459
2021-07-16 10:17:11.053  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Fetching Scan data for Id 1020766
2021-07-16 10:17:11.063  INFO 125534 --- [      flow-web1] c.c.s.s.CxService                         [H9csmv7i] : Fetching custom fields from project ID 188
2021-07-16 10:17:11.082 ERROR 125534 --- [      flow-web1] c.c.f.c.FlowAsyncConfig                   [H9csmv7i] : Error has occurred calling initiateAutomation. MappingException: ModelMapper mapping errors:

1) Error mapping com.checkmarx.sdk.dto.ScanResults to com.checkmarx.sdk.dto.ScanResults

1 error
2021-07-16 10:17:11.082 ERROR 125534 --- [      flow-web1] c.c.f.c.FlowAsyncConfig                   [H9csmv7i] : Parameter value - ScanRequest(namespace=<namespace>, application=<repo_name>, org=null, team=<team_name>, project=<project_name>, cxFields=null, site=null, repoUrl=<repo_url>, repoName=<repo_name>, branch=master, mergeTargetBranch=null, mergeNoteUri=null, repoProjectId=null, refs=refs/heads/master, email=[<emails>], incremental=false, scanPreset=<preset_name>, excludeFiles=[*_test.go, *.twirp.go], excludeFolders=[vendor, tests, integrationtests, __test__, api-mocks], repoType=GITHUB, product=CX, bugTracker=BugTracker(type=JIRA, customBean=null, projectKey=<project_key>, issueType=Vulnerability, openTransition=Open, closeTransition=Closed, assignee=null, openStatus=[Open, Ready For Dev, In Development, In Code Review, Waiting For QA, In QA, Waiting For Approval, Waiting For Merge, Research, In Review, Blocked], closedStatus=[Closed, Cancelled], closeTransitionField=resolution, closeTransitionValue=Done, priorities={High=High, Medium=Medium, Low=Low, Informational=Trivial, X=Y}, fields=[com.checkmarx.flow.dto.Field@56001dc5, com.checkmarx.flow.dto.Field@208be996]), type=null, activeBranches=[master], filter=FilterConfiguration(sastFilters=EngineFilterConfiguration(simpleFilters=[Filter(type=SEVERITY, value=High)], scriptedFilter=com.checkmarx.sdk.dto.filtering.ScriptedFilter@6cb693eb), scaFilters=EngineFilterConfiguration(simpleFilters=[], scriptedFilter=null)))
2021-07-16 10:17:11.082 ERROR 125534 --- [      flow-web1] c.c.f.c.FlowAsyncConfig                   [H9csmv7i] : Error details:

org.modelmapper.MappingException: ModelMapper mapping errors:

1) Error mapping com.checkmarx.sdk.dto.ScanResults to com.checkmarx.sdk.dto.ScanResults

1 error
	at org.modelmapper.internal.Errors.throwMappingExceptionIfErrorsExist(Errors.java:380)
	at org.modelmapper.internal.MappingEngineImpl.map(MappingEngineImpl.java:80)
	at org.modelmapper.ModelMapper.mapInternal(ModelMapper.java:573)
	at org.modelmapper.ModelMapper.map(ModelMapper.java:447)
	at com.checkmarx.sdk.dto.ScanResults.mergeWith(ScanResults.java:202)
	at com.checkmarx.flow.service.FlowService.lambda$runScanRequest$3(FlowService.java:62)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at com.checkmarx.flow.service.FlowService.runScanRequest(FlowService.java:60)
	at com.checkmarx.flow.service.FlowService.initiateAutomation(FlowService.java:39)
	at com.checkmarx.flow.service.FlowService$$FastClassBySpringCGLIB$$3503db27.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:771)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:56)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:175)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:95)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749)
	at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassCastException: null`

### Environment Details

OS: Ubuntu 20.04.2 LTS
Java: openjdk 11.0.11 2021-04-2

**Comment by AvivCx** [comment]
Hi @y4ppieflu . thanks for raising this ticket
1. do you know the steps to reproduce this issue with our Jira project?
2.  can you contact me or @jbrotsos and send debug level logs?
Thank you

**Comment by y4ppieflu** [comment]
Hi @AvivCx 
I think that I already have debug level logs enabled in application.yml and the provided excerpt is everything that gets logged. 
I can also provide application.yml and cx.config with per-repo overrides if it helps.
Thanks

**Comment by AvivCx** [comment]
@y4ppieflu 
yes, it can help. please remove every secret or sensitive data from the files before you share it
Thanks

**Comment by y4ppieflu** [comment]
@AvivCx 
Sure, here you go:
- application.yml
```
server:
  port: ${PORT:80}

logging:
  file: /opt/cxflow/cx-flow.log
  level:
    com:
      checkmarx:
        flow:
          service: TRACE
    org:
      apache:
        http:
          wire: INFO
      springframework:
        web:
          client:
            RestTemplate: DEBUG

cx-flow:
  contact: <email>
  token: <token>
  bug-tracker: JIRA
  branches:
    - master
  filter-severity:
    - High
  mitre-url: https://cwe.mitre.org/data/definitions/%s.html
  wiki-url: https://checkmarx.atlassian.net/wiki/spaces/AS/pages/79462432/Remediation+Guidance
  codebash-url: https://cxa.codebashing.com/courses/
  mail:
     host: <mail_relay>
     port: 25
     username: <user>
     password: <pass>
     enabled: true    

checkmarx:
  version: 9.0
  username: <username>
  password: <password>
  client-id: resource_owner_client
  client-secret: <client_secret>
  scope: access_control_api sast_rest_api
  base-url: https://hostname
  multi-tenant: true
  configuration: Multi-language Scan
  scan-preset: Checkmarx Default
  team: <team>
  url: ${checkmarx.base-url}/cxrestapi
  preserve-xml: true
  incremental: false
  incremental-num-scans: 10
  jira-project-field: jira-project
  jira-issuetype-field: jira-issuetype
  jira-custom-field: jira-fields
  jira-assignee-field: jira-assignee
  portal-url: ${checkmarx.base-url}/cxwebinterface/Portal/CxWebService.asmx
  exclude-files: "<file_list>"
  exclude-folders: "<folder_list>"

jira:
  url: https://jira.hostname
  username: <username>
  token: <token>
  project: <project>
  issue-type: Vulnerability
  priorities:
    High: High
    Medium: Medium
    Low: Low
    Informational: Trivial
  open-transition: Open
  close-transition: Closed
  close-transition-field: resolution
  close-transition-value: Done
  close-false-positive-transition-value: Won't Do
  open-status:
    - Open
    - Ready For Dev
    - In Development
    - In Code Review
    - Waiting For QA
    - In QA
    - Waiting For Approval
    - Waiting For Merge
    - Research
    - In Review
    - Blocked
  closed-status:
    - Closed
    - Cancelled

github:
  webhook-token: <token>
  token: <token>
  url: https://github.com/
  api-url: https://api.github.com/repos/
  false-positive-label: false-positive
  block-merge: false
  error-merge: true
  cx-summary: true
```

- cx.config
```
{
    "version": 1.0,
    "team": "<team>",
    "project": "${repo}",
    "sast": {
        "preset": "<preset>",
        "engineConfiguration": "Multi-language Scan"
    },
    "additionalProperties": {
        "cxFlow": {
            "branches": ["master"],
            "bugTracker":"JIRA",
            "jira": {
                "fields": [
                    {
                        "type": "static",
                        "name": "Domain",
                        "jira_field_name": "Domain",
                        "jira_field_type": "single-select",
                        "jira_default_value": "<domain>"
                    },
                    {
                        "type": "static",
                        "name": "Team",
                        "jira_field_name": "Team",
                        "jira_field_type": "single-select",
                        "jira_default_value": "<team>"
                    }
                ]
            }
        }
    }
}
```

**Comment by AvivCx** [comment]
Hi @y4ppieflu 
you mentioned cxflow 'occasionally' fail to parse.... did you notice any pattern, or specific configuration that causing this exception?
Can you give simple steps to reproduce?

Thanks

**Comment by y4ppieflu** [comment]
Hi @AvivCx 
It's hard to say exactly - we have around 200 projects that are getting scanned per each new pull request to master. I see this particular error happens several times a day and I need to restart the CxFlow service in order to get it back to normal mode (otherwise it is not sending scan results comments to GitHub and is not creating/updating Jira tickets). 
My assumption is that this may be caused by failing scans on Checkmarx side (e.g. in Checkmarx console I see failed scans with "No engine can reserve memory for this scan" message), but I am not 100% sure.

**Comment by y4ppieflu** [comment]
Hi @AvivCx 
Probably caused by this bug in modelmapper - https://github.com/modelmapper/modelmapper/issues/606.
For now I have downgraded openjdk to 11.0.7 to check if it resolves the issue.

**Comment by AvivCx** [comment]
Thanks for this information @y4ppieflu!
Hi @jbrotsos 
Let's add this to the pool for the coming sprints

**Comment by umeshwaghode** [comment]
@y4ppieflu Did it reproduce anytime after you downgraded OpenJDK to 11.0.7?

**Comment by y4ppieflu** [comment]
@umeshwaghode Nope, CxFlow works absolutely stable since downgrade.

**Comment by jbrotsos** [comment]
Released in 1.6.33
